main/core/src/eval/Evaluator.scala

```diff
@@ -362,15 +382,302 @@ case class Evaluator(home: os.Path,
+
+    // The scheduled and not yet finished futures (Java!)
+    var futures = List[(java.util.concurrent.Future[FutureResult], TerminalGroup)]()
+
+    try {
+
+      val interGroupDeps: Map[TerminalGroup, Seq[TerminalGroup]] = findInterGroupDeps(sortedGroups)
+      evalLog.debug(s"${interGroupDeps} (took ${System.currentTimeMillis() - startTime} msec)")
+
+      // State holders, only written to from same thread
+      // The unprocessed terminal groups
+      var work = sortedGroups.items().toList
+      // The currently scheduled (maybe not started yet) terminal groups
+      var inProgress = List[TerminalGroup]()
+      // The finished terminal groups
+      var doneMap = Map[TerminalGroup, Boolean]().withDefaultValue(false)
+      // The fact that at least one task failed
+      @volatile var someTaskFailed: Boolean = false
+
+      if (work.size != work.distinct.size) {
+        evalLog.error(s"Work list contains ${work.distinct.size - work.size} duplicates!")
+      }
+
+//      evalLog.debug(s"Work: ${work.size} -- ${work.map(t => printTerm(t._1)).mkString(", ")}")
+
+      /**
+        * Checks for terminal groups, that have no unresolved dependencies and schedule them to run via the executor service.
+        */
+      def scheduleWork(issuer: String): Unit = {
+        // early exit
+        if (work.isEmpty || inProgress.size > effectiveThreadCount) return
+
+        val scheduleStart = System.currentTimeMillis()
+
+        // newInProgress: the terminal groups without unresolved dependencies
+        // newWork: the terminal groups, with unresolved dependencies (need to wait longer)
+        val (newInProgress, newWork) = work.partition { termGroup =>
+          val deps = interGroupDeps(termGroup)
+          deps.isEmpty || deps.forall(d => doneMap(d))
+        }
+
+        // update state
+        work = newWork
+        inProgress = inProgress ++ newInProgress
+
+        evalLog.debug(s"Search for ${newInProgress.size} new dep-free tasks took ${System.currentTimeMillis() - scheduleStart} msec")
+
+        if (!newInProgress.isEmpty) {
+          evalLog.debug(s"Scheduling ${newInProgress.size} new tasks (issuer: ${issuer}): ${newInProgress.map(t => printTerm(t._1)).mkString(", ")}")
+
+          // schedule for parallel execution
+          newInProgress.zipWithIndex.foreach {
+            case (curWork @ (terminal, group), index) =>
+
+              //              val missingDependencies = group.indexed.flatMap(_.inputs).filter(t => !results.contains(t) && !group.contains(t))
+              //              if (!missingDependencies.isEmpty) {
+              //                evalLog.error(s"Missing resolved dependencies for terminal group: ${printTerm(terminal)}\n  ${missingDependencies.mkString(",\n  ")}")
+              //              }
+
+              val workerFut: java.util.concurrent.Future[FutureResult] = completionService.submit { () =>
```

Comment:
It's a bit non-obvious from skimming the code that this block of code multithreaded, while the rest of the code in this method is single threaded and runs only on the main thread. I read over this entire method 4-5 times before realizing that fact

Could we move this "parallel" code into a separate method or class, so it's clearly separated from the single threaded portion of `def evaluateParallel`?

Reply:
Sounds like a good idea. But as a result I think that means I need to move the whole code into a separate class, to better encapsulate shared state.

Reply:
That sounds reasonable. In general, I think this entire `evaluateParallel` method is complex enough to be a standalone class, with the logic broken down into other smaller classes or methods. You'll have to use your judgement, but I think having a huge complex method in an already complex `Evaluator.scala` file is a bit too much

Reply:
I started with an inner class and splitted into some vals and defs as a first iteration.